src/components/utils/include/utils/timer.h

```diff
@@ -84,47 +90,127 @@ class Timer {
+   *
    */
-  void OnTimeout();
+  virtual void UpdateTimeOut(const uint32_t timeout_milliseconds);
 
   /**
-   * @brief method for setting correct timeout.
-   * @param timeout - if it`s value = 0, timeout will be setted to 1
-   * There would be no way to stop thread if timeout in lopper will be 0
-   * and if we puts to timer_create zero timeout then we get sys error(22)
+   * @brief GetTimeout
+   * @return returns timeout
    */
-  void SetTimeoutUnsafe(const Milliseconds timeout);
+  Milliseconds GetTimeout() const;
 
+ private:
   /**
-   * @brief startUnsafe, stopUnsafe - methods used for correct synchronization
-   *  and must be used only with sync_primitive (auto_lock e.g.)
-   * @return true if start/stop successfull, false if unsuccessfull
+   * @brief Delegate release timer, will call callback function one time
+   * or call delegate every timeout function while stop()
+   * won't be called. It's depend on flag.
    */
-  void StartUnsafe();
-  bool StopUnsafe();
+  class TimerDelegate : public threads::ThreadDelegate {
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     */
+    explicit TimerDelegate(Timer* timer);
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~TimerDelegate();
+
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+    /**
+     * @brief Called by thread::thread to free all allocated resources.
+     */
+    void exitThreadMain() OVERRIDE;
+
+    /**
+     * @brief Set new Timeout
+     * @param timeout_milliseconds New timeout to be set
+     */
+    virtual void SetTimeOut(const uint32_t timeout_milliseconds);
+
+    /**
+      * @brief Wait until timer will start
+      */
+    virtual void WaitUntilStart();
+
+    /**
+     * @brief Quits threadMain function after next loop.
+     */
+    virtual void ShouldBeStopped();
+
+    /**
+     * @brief Restarts non-loop timer after current iteration.
+     */
+    virtual void ShouldBeRestarted();
+
+    /**
+     * @brief Return flag IsGoingStop
+     */
+    bool IsGoingToStop() const;
+
+    /**
+     * @brief Gets timeout with overflow check
+     * @return timeout
+     */
+    inline int32_t get_timeout() const {
+      return std::min(
+          static_cast<uint32_t>(std::numeric_limits<int32_t>::max()),
+          timeout_milliseconds_);
+    }
+
+    /**
+      * @brief Make class delegate repeatable
+      */
+    inline void make_repeatable() {
+      is_repeatable_ = true;
+    }
+
+   protected:
+    Timer* timer_;
+    uint32_t timeout_milliseconds_;
+    sync_primitives::Lock state_lock_;
+    sync_primitives::ConditionalVariable termination_condition_;
+    sync_primitives::ConditionalVariable starting_condition_;
+    sync_primitives::atomic_bool stop_flag_;
+    sync_primitives::atomic_bool restart_flag_;
+    sync_primitives::atomic_bool is_started_flag_;
+    sync_primitives::atomic_bool is_repeatable_;
```

Comment:
@anosach-luxoft , because `state_lock_` preventing call at the same time timer delete, but flags have to be independent from it adn have to be thread safe. That is why, I won't do it.